decisions/0003-package-for-shared-server-features.md

```diff
@@ -0,0 +1,111 @@
+- reduce bugs from `remix` internal complexity
+
+### Non-goals
+
+- maintainability of `remix` repo
+- more portable `app/` directory
+- reduce or increase number of packages
+
+Some of these are nice-to-haves, but not the primary reasons for this proposal.
+
+### Issues
+
+1. Users can import features from their runtime that aren't supported by their adapter:
+
+```tsx
+import { createFileSessionStorage } from "@remix-run/node";
+// Implementation only depends on server runtime, but _availability_ depends on the adapter!
+// ❌ This will crash in netlify nor in vercel!
+```
+
+2. Docs, examples, and videos refer to `@remix-run/node` by default for shared functionality like `json`:
+
+```tsx
+import { json } from "@remix-run/node";
+// Need docs to tell users that they should import from their corresponding runtime package, 
+// and not from `@remix-run/server-runtime`.
+```
+
+3. Keeping shared exports consistent across runtime packages is [error-prone](https://github.com/remix-run/remix/pull/3215) and tedious.
+
+## Decision
+
+Users should import features from _adapter packages_, not from runtime packages. Each adapter can export only the features it supports. Shared features will be imported from a new package, `@remix-run/server`.
+
+This means there won't be any imports needed from the runtime packages (`@remix-run/{cloudflare,deno,node`}) and these will become an implementation detail for most users. Only developers that need to develop adapters will import from the runtime packages directly.
+
+### User-facing packages
+
+Users still need to import from 4 Remix packages:
+
+- Dev `@remix-run/dev` (unchanged)
+- Shared server features `@remix-run/server`
+- Adapter-specific server features (`@remix-run/{arc,cloudflare-pages,cloudflare-workers,express,fly,netlify,vercel}`)
+- Renderer `@remix-run/react` (unchanged)
+
+User code would look like this:
+```tsx
+// server
+import { json } from "@remix-run/server";
+// adapter
+import { createRequestHandler, createFileSessionStorage } from "@remix-run/fly"
+// renderer
+import { useLoaderData } from "@remix-run/react";
+```
+
+### Internal packages
+
+- Runtime packages for authoring adapters (`@remix-run/{cloudflare,deno,node}`)
+- Package for authoring new runtimes (`@remix-run/server-runtime`), only including factory functions like:
+  - `createCookieFactory`
+  - `createSessionStorageFactory`
+  - `createCookieSessionStorageFactory`
+  - `createMemorySessionStorageFactory`
+
+### Migration
+
+Include a new migration for updating imports. Should be a straightforward copy of the existing [replace-remix-imports](https://github.com/remix-run/remix/blob/main/packages/remix-dev/cli/migrate/migrations/replace-remix-imports/index.ts) migration with [small tweaks to the import maps](https://github.com/remix-run/remix/blob/main/packages/remix-dev/cli/migrate/migrations/replace-remix-imports/transform/mapNormalizedImports/packageExports.ts).
```

Comment:
We could use the `replace-remix-imports` migration for this instead of copy/pasting the whole thing over.

Only things needed to change are:
- The imports map (as already mentioned)
- The imports to change
  Currently we're only looking at `remix` imports. This should be changed to also include `@remix-run/*` packages.